### PR TITLE
fixing putHandler

### PR DESCRIPTION
**Pull request by cryptix**
trying to finally get a working `putHandler` for the writable http gateway.
 
License: MIT
Signed-off-by: Henry <cryptix@riseup.net>

### CODE REVIEW

**Comment by cryptix** [comment]
Since #1845 is based on 0.4, I tried my best to get the `putHandler` back to work.

I fixed overwriting an existing path and 'object add-link' but I'm not sure how in line this are with the old sharness tests and or if we still wont their behavior.

**Comment by cryptix** [comment]
cc @jbenet @whyrusleeping

File: core/corehttp/gateway_handler.go [hunk 1]
```diff
@@ -273,116 +276,90 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
 
-	tctx, cancel := context.WithTimeout(ctx, time.Minute)
-	defer cancel()
-	// TODO(cryptix): could this be core.Resolve() too?
-	rootnd, err := i.node.Resolver.DAG.Get(tctx, key.Key(h))
-	if err != nil {
-		webError(w, "Could not resolve root object", err, http.StatusBadRequest)
-		return
-	}
+		createfunc := func() *dag.Node {
```
Comment (rht) [inline core/corehttp/gateway_handler.go]: This is `uio.NewEmptyDirectory`

**Comment by rht** [comment]
Since netstat is not available in the test env, the first test should be commented out.

File: core/corehttp/gateway_handler.go [hunk 2]
```diff
@@ -273,116 +275,87 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
 }
 
-func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request) {
-	newnode := uio.NewEmptyDirectory()
-
-	key, err := i.node.DAG.Add(newnode)
-	if err != nil {
-		webError(w, "Could not recursively add new node", err, http.StatusInternalServerError)
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
+	rootPath := path.Path(r.URL.Path)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: could probably use `path.Parse` here

File: core/corehttp/gateway_handler.go [hunk 3]
```diff
@@ -273,116 +275,87 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
 		return
 	}
 
 	var newnode *dag.Node
-	if pathext[len(pathext)-1] == '/' {
+	if rsegs[len(rsegs)-1] == "QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn" {
 		newnode = uio.NewEmptyDirectory()
 	} else {
+		var err error
 		newnode, err = i.newDagFromReader(r.Body)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: in these cases i prefer explicit rebinding as opposed to something that might lead to accidental shadowing:
```
dfr, err := i.newDagFromReader(r.Body)
if err != nil {
....
}

newnode = dfr
```

I personally find it cleaner and easier to reason about.

File: core/corehttp/gateway_handler.go (hunk 1 above)
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: `createfunc := uio.NewEmptyDirectory`

File: core/corehttp/gateway_handler.go [hunk 4]
```diff
@@ -273,116 +275,84 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
-	defer cancel()
-
-	ipfsNode, err := core.Resolve(ctx, i.node, path.Path(urlPath))
-	if err != nil {
-		// FIXME HTTP error code
-		webError(w, "Could not resolve name", err, http.StatusInternalServerError)
-		return
+	var newPath string
+	if len(rsegs) > 1 {
+		newPath = filepath.Join(rsegs[2:]...)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: we dont want to use filepath here, on windows systems this would put forward slashes in the ipfspaths, which i dont think works like we expect it to.

File: core/corehttp/gateway_handler.go [hunk 5]
```diff
@@ -273,116 +275,84 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
 
-	tctx, cancel := context.WithTimeout(ctx, time.Minute)
-	defer cancel()
-	// TODO(cryptix): could this be core.Resolve() too?
-	rootnd, err := i.node.Resolver.DAG.Get(tctx, key.Key(h))
-	if err != nil {
-		webError(w, "Could not resolve root object", err, http.StatusBadRequest)
-		return
-	}
+		newPath := filepath.Join(rsegs[2:]...)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: same filepath comment as above.

File: core/corehttp/gateway_handler.go [hunk 6]
```diff
@@ -273,116 +275,84 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
 
-	// resolving path components into merkledag nodes. if a component does not
-	// resolve, create empty directories (which will be linked and populated below.)
-	pathNodes, err := i.node.Resolver.ResolveLinks(tctx, rootnd, components[:len(components)-1])
-	if _, ok := err.(path.ErrNoLink); ok {
-		// Create empty directories, links will be made further down the code
-		for len(pathNodes) < len(components) {
-			pathNodes = append(pathNodes, uio.NewDirectory(i.node.DAG).GetNode())
+		e := dagutils.NewDagEditor(i.node.DAG, rnode)
+		err = e.InsertNodeAtPath(i.node.Context(), newPath, newnode, uio.NewEmptyDirectory)
```
Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: we should probably derive our own context from the nodes context at the top of the handler, that way we can be sure to cancel everything afterwards

File: path/resolver.go [hunk 7]
```diff
@@ -22,12 +22,12 @@ var ErrNoComponents = errors.New(
 
 // ErrNoLink is returned when a link is not found in a path
 type ErrNoLink struct {
-	name string
-	node mh.Multihash
+	Name string
```
Comment (whyrusleeping) [inline path/resolver.go]: :+1: to introspectable errors

File: core/corehttp/gateway_handler.go (hunk 4 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: ugh, true.. I always manage to ignore that.

File: core/corehttp/gateway_handler.go (hunk 6 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: sure - btw: whats the status of `http.CloseNotifier`?

Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: oh crap, we can do that now.

Comment (cryptix) [inline core/corehttp/gateway_handler.go]: Awesome! should I do it in this PR? feels like..

Comment (whyrusleeping) [inline core/corehttp/gateway_handler.go]: yeah.... lets put that in another PR

**Comment by whyrusleeping** [comment]
just a few comments, then squash some commits so none of them break. (i think some of the earlier ones break) then LGTM

File: core/corehttp/gateway_handler.go (hunk 6 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: Okay, leaving this one be than for now

**Comment by cryptix** [comment]
CR adressed and squashed.

**Comment by ion1** [comment]
* [RFC 7231: POST, PUT, DELETE](https://tools.ietf.org/html/rfc7231#section-4.3.3)
* [RFC 6266: Content-Disposition/filename](https://tools.ietf.org/html/rfc6266#section-4.3)

Please correct me if I have misunderstood the RFC text, but it seems to me like the methods should have the following semantics in this context:

**`POST /ipfs`** creates a new resource under `/ipfs` whose name (in this case: hash) is determined by the gateway and receives the response:

```
303 See Other
Location: /ipfs/QmFoo
```

**`POST /ipfs/QmBar` or `POST /ipfs/QmBar/baz`** request a resource to be added *under* `QmBar` and `QmBar/baz` respectively. The server has to choose the name of the resource; we should dictate that the request must have a `Content-Disposition: …; filename="…"` header and use that for the name.

Technically the resulting resource will be added under a new `/ipfs/QmQuux` which is the result of patching `QmBar` but this might not be too great a sin. The response to `POST /ipfs/QmBar/baz` with `filename="my-file"`:

```
303 See Other
Location: /ipfs/QmQuux/baz/my-file
```

We will have to decide what to do if the request would overwrite an existing file in `QmBar`. The choices:

* Just overwrite it and give a `303 See Other` response as above. Any qualms about `POST` apparently deleting content may be soothed by realizing that the original `QmBar` is still there with the original file. This would be my choice.
* Give a `409 Conflict` response and require the client to `DELETE` the original file first. This increases complexity, requires two round trips and results in a useless intermediate object being created.

**`PUT /ipfs`** is not meaningful (“I expect a future `GET /ipfs` to return this content”). `405 Method Not Allowed`.

**`PUT /ipfs/QmFoo`** can only succeed if `QmFoo` is in fact the hash of the object being uploaded (requiring the client to compute the hash in advance). Either don’t support this and respond with `405 Method Not Allowed` or check whether the hash matches and respond with `201 Created` or `409 Conflict`.

**`PUT /ipfs/QmBar/baz`** is not meaningful: the gateway might not know anything about `QmBar` under which `baz` is requested to be created, and if it does know, `QmBar` already exists and is immutable. `405 Method Not Allowed`.

This can not be used to patch `QmBar`: the RFC dictates that the server “MUST” respond to a successful `PUT` with `2xx`. A `303 See Other` response after creating a patched object would not be valid behavior.

**`DELETE /ipfs`** is not meaningful. `405 Method Not Allowed`.

**`DELETE /ipfs/QmFoo`** is not meaningful. `405 Method Not Allowed`.

**`DELETE /ipfs/QmBar/baz/my-file`** can be used to patch `QmBar` to remove `my-file`. The RFC only says `DELETE` “SHOULD” give a `2xx` response. We should not be in violation if we create a patched object and respond with:

```
303 See Other
Location: /ipfs/QmQuux/baz
```

**Comment by rht** [comment]
You're suggesting PUT should be used only for mutable (ipns) case?

**Comment by ion1** [comment]
`PUT` could be used with IPNS, yes. The rule seems to be:

After a successful `PUT <location>`, a subsequent `GET <location>` will return what you put.

This implies mutation.

**Comment by mildred** [comment]
`UpdateNodeLink` in merkledag/node.go can probably be removed as part of this PR. gateway_handler was the only one using it.

Also, I'm generally liking PUT much more than POST when it can be used, but I agree that in the case of ipfs, resources are immutable and PUT doesn't makes sense except at the root when the client already knows the hash of the node it uploads. POST with Content-Disposition should probably be implemented instead.

File: core/corehttp/gateway_handler.go [hunk 8]
```diff
@@ -273,116 +275,82 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
+		log.Warningf("putHandler: unhandled resolve error %T", ev)
+		webError(w, "could not resolve root DAG", ev, http.StatusInternalServerError)
 		return
 	}
 
 	i.addUserHeaders(w) // ok, _now_ write user's headers.
-	w.Header().Set("IPFS-Hash", key.String())
-	http.Redirect(w, r, ipfsPathPrefix+key.String()+"/"+strings.Join(components, "/"), http.StatusCreated)
+	w.Header().Set("IPFS-Hash", newkey.String())
+	http.Redirect(w, r, filepath.Join(ipfsPathPrefix, newkey.String(), newPath), http.StatusCreated)
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: path

File: test/sharness/t0111-gateway-writeable.sh [hunk 9]
```diff
@@ -14,9 +14,10 @@ test_launch_ipfs_daemon
 
 port=$PORT_GWAY
 
-test_expect_success "ipfs daemon listening to TCP port $port" '
-  test_wait_open_tcp_port_10_sec "$PORT_GWAY"
-'
+# TODO(cryptix): netstat not avail on testing system?
+#test_expect_success "ipfs daemon listening to TCP port $port" '
+#  test_wait_open_tcp_port_10_sec "$PORT_GWAY"
+#'
```
Comment (jbenet) [inline test/sharness/t0111-gateway-writeable.sh]: can probably use the pollEndpoint thing

**Comment by jbenet** [comment]
The PR is hard to follow because changes change code quite a bit. i haven't read it all. but if LGTWhyrusleeping, and the tests pass, i'd be ok to merge it.

(in the end i want to split out all the gateway stuff into own repo)

---

i think ion covered it very well, all that SGTM.

**Comment by cryptix** [comment]
@jbenet it's easier to follow if you just look at the new code.

Do we want to merge it now and do the change to POST in a follow up PR? The changes will only get bigger because the tests need to change as well. OTOH we would introduce features again that aren't there to stay.. 

I'd also like to introduce pure go tests for the things @ion1 brought up. I think it's easier to outline and test all those in a table based manner.

**Comment by whyrusleeping** [comment]
I would be comfortable merging this as its something that is disabled by default.

**Comment by cryptix** [comment]
rebased.

I'll start a new branch based on this one for the POST conversion, if we want to do it in this PR I can still push it here.

File: core/corehttp/gateway_handler.go [hunk 10]
```diff
@@ -273,116 +275,82 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
-	defer cancel()
-	// TODO(cryptix): could this be core.Resolve() too?
-	rootnd, err := i.node.Resolver.DAG.Get(tctx, key.Key(h))
-	if err != nil {
-		webError(w, "Could not resolve root object", err, http.StatusBadRequest)
-		return
-	}
+	var newkey key.Key
+	rnode, err := core.Resolve(i.node.Context(), i.node, rootPath)
+	switch ev := err.(type) {
```
Comment (rht) [inline core/corehttp/gateway_handler.go]: This is a switch from err casting.
This is cleaner, and should be used across the code.

In current master, where cast is used:
```shell
$ grep -rE 'err\.\(' . | grep -v Godeps
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./routing/dht/ext_test.go:		if merr, ok := err.(u.MultiErr); ok && len(merr) > 0 {
./core/corehttp/gateway_handler.go:	if _, ok := err.(path.ErrNoLink); ok {
./core/corehttp/gateway_handler.go:	if _, ok := err.(path.ErrNoLink); ok {
./core/commands/add.go:		if _, ok := err.(*hiddenFileError); ok {
./p2p/net/conn/dial_test.go:			if ne, ok := err.(net.Error); ok && ne.Temporary() {
./p2p/net/conn/dial.go:	if nerr, ok := err.(net.Error); ok && nerr.Timeout() {
./p2p/net/conn/dial.go:	errno, ok := err.(syscall.Errno)
./cmd/ipfs/main.go:	switch e := err.(type) {
```

**Comment by rht** [comment]
Unless I'm missing something, to implement POST for composite files on 0.3.x, doesn't it mean the code for `ipfs add` is duplicated almost verbatim?

**Comment by rht** [comment]
How is @ion1's `PUT /ipfs/QmBar/baz` case addressed in test/sharness/t0111-gateway-writeable.sh#L60 ? The new hash isn't precomputed.

File: core/corehttp/gateway_handler.go [hunk 11]
```diff
@@ -273,116 +275,82 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
 	http.Redirect(w, r, ipfsPathPrefix+k.String(), http.StatusCreated)
 }
 
-func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: what happened to this part? do empty dirs still work?

File: core/corehttp/gateway_handler.go [hunk 12]
```diff
@@ -273,116 +275,82 @@ func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
...
-	newnode := uio.NewEmptyDirectory()
-
-	key, err := i.node.DAG.Add(newnode)
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request) {
```
Comment (jbenet) [inline core/corehttp/gateway_handler.go]: should derive a ctx here and cancel it to ensure everything is wiped at the end.

```go
ctx, cancel := context.WithCancel(i.node.Context())
defer cancel()
```

**Comment by jbenet** [comment]
@cryptix a lot of people want this. eta on addressing CR?

File: core/corehttp/gateway_handler.go (hunk 11 above)
Comment (cryptix) [inline core/corehttp/gateway_handler.go]: yup - works by virtue of core.Resolve and friends iirc. There is also a sharness test that checks against the hash.

**Comment by cryptix** [comment]
@jbenet done :)

**Comment by jbenet** [comment]
Thanks @cryptix ! LGTM. I'll merge this for now -- so @victorbjelkholm can use in IPFSBin and @cloutier  and @didiercf can use in ipfspics. we can move from `PUT` to `POST` in the future.

**Comment by rht** [comment]
And [ipfs-wiki](https://github.com/jamescarlyle/ipfs-wiki), which uses `python -m SimpleHTTPServer`.

**Comment by jbenet** [comment]
cc @jamescarlyle o/
